Auto-open project owner of the opened file.

sdedic opened the pull request:
Especially with gradle subproject, it happens that if the gradle project is not opened, diagnostics run on the project's sources report missing identifiers. The LSP server can internally automatically open project owner when the LSP client opens a file from that project.

This PR performs `asyncOpenSelectedProjects` using the file being opened. If the project is *being opened* (including priming build), the actual `OpenedProjects.open` call will be postponed until after that process completes.
For projects that whose open operation completed already, the Future in `beingOpened` is completed, so calling `asyncOpenSelectedProjects` on an opened project will also complete immediately.

I've changed the semantic of `beingOpened` to keep the completed Futures as a sign that a priming build is not needed (again).

JaroslavTulach commented on java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/TextDocumentServiceImpl.java:
```diff
@@ -1784,7 +1784,11 @@ public void didOpen(DidOpenTextDocumentParams params) {
             openedDocuments.put(params.getTextDocument().getUri(), doc);
-            runDiagnoticTasks(params.getTextDocument().getUri());
+            
+            // attempt to open the directly owning project, delay diagnostics after project open:
```
Possibly this could open a completely unrelated project - will the user be asked for a consent to trust the project in that case?

Btw. there is [NETBEANS-5204](https://github.com/apache/netbeans/pull/2629/) to help you find whether a subproject is related to already opened project or not.

sdedic replied:
Good point with the project relationships; will try to add checks.

JaroslavTulach reviewed (approved):
Could this be done for NetBeans as well? It would be great to see a bubble in top right corner asking _"Hey, you are working on a file without opened project. Open xyz?"_

sdedic commented:
> Could this be done for NetBeans as well? It would be great to see a bubble in top right corner asking "File xyz belongs to project abc. Open it and analyze it?"

Yes, this would be nice. In NetBeans we have `EditorRegistry`, so Projects UI could eventually track the opened files and ask as you suggested. 

Sadly this must be duplicated (somehow) for LSP server, as we do not have some abstraction for an actively-worked-on file except the EditorRegistry that is tied to Swing.

Tracked as  [NETBEANS-5454](https://issues.apache.org/jira/browse/NETBEANS-5454)

sdedic commented:
Suspending. To work with Gradle projects, I need to implement [NETBEANS-5430](https://issues.apache.org/jira/browse/NETBEANS-5430) first. Maven projects, which already support priming, lacks the support for project hirerarchy queries -- [NETBEANS-5455](https://issues.apache.org/jira/browse/NETBEANS-5455)

sdedic commented:
Rebased on master to get Maven project parent and root queries to the branch.

I have added the user questions when a project is implicitly opened. The question is asked only once per LSP server instance, so if the user rejects, he won't be bothered again for the project, but the diagnostics etc won't be working much (but the user wanted so).
The question is aked if **the project and no of its parent projects** are already opened **in the workspace**. So if the user e.g. opens a folder with some parent in VSCode, the subprojects (as they are contained in + built with the parent) are opened without questions.

JaroslavTulach commented on java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/LspServerState.java:
```diff
@@ -50,6 +51,23 @@
     public CompletableFuture<Project[]>   asyncOpenSelectedProjects(List<FileObject> fileCandidates);
     
+    /**
+     * Opens project oo behalf of a file. This makes the project 'second-class citizen' in LSP: it will be
```
Typo `oo`

sdedic commented:
Squashed + typo fixed. Waiting for build to proceed before merge.

sdedic commented:
Windows & Mac builds have connectitivy issues. Travis build  for Graalvm (latest) as well. The only 'normally' failing test is `Test java.hints batch1 without nb-javac on Java 14` - but not from this PR.
